### ISSUE TITLE
Standardize on factory facet?

### ISSUE THREAD

**Issue by mindplay-dk** [container-interop/container-interop#44]
Can we standardize on the factory facet implemented by many containers?

Looking at all of the following method signatures, they are practically identical:

  * `make($name, array $parameters = [])` in [PHP-DI](https://github.com/PHP-DI/PHP-DI/blob/master/src/DI/Container.php#L151)
  * `make($name, array $args = array())` in [Auryn](https://github.com/rdlowrey/auryn/blob/master/lib/Injector.php#L335)
  * `create($class_name, $map = array())` in [Unbox](https://github.com/mindplay-dk/unbox/blob/master/src/Container.php#L369)
  * `build($concrete, array $parameters = [])` in [Laravel](https://github.com/illuminate/container/blob/master/Container.php#L723)

Basically every container offers this functionality with basically the same interface and behavior, so this should lend itself to standardization pretty naturally?

**Comment by mnapoli** [comment]
I'm more or less +1 I personally don't make extensive use of that feature (unlike `invoke()`), but if it's used in some frameworks or libraries and the interface would allow to decouple those from containers then why not.

**Comment by mindplay-dk** [comment]
> personally don't make extensive use of that feature (unlike `invoke()`)

Yes, arguably it's much less frequently used - I don't think that means it's less important though?

On a related note - some containers do not distinguish between factory and singleton components; that is, some containers, such as Pimple, have a single API method, which might return either a singleton component or a new instance on each call, depending on how it was configured.

I tried that myself in earlier attempts to make a DI container, and decided I didn't like it much. A service container to me is something very different from a factory - I suppose some people might think it's great that they can switch between singleton or factory at the configuration level, though I have a hard time thinking up with a case where it doesn't matter to the consumer if it's getting a new or singleton instance... How do you guys feel about that?

Obviously these containers can't support the factory interface, right?

Based on that behavior, could they be said to support the primary `ContainerInterface`, even if they were to implement it? Does the RFC prescribe singleton behavior for that interface?

**Comment by drealecs** [comment]
Let me share you why I would give it a +1:

In case of old code that you can't fix over night, old code that might not have stateless services and those services having state injected in constructor, sometimes the easy interface fix in order to create some tests (required for further refactoring) would mean replacing **new Service($state)** call with a call to an injected ServiceFactory: **$this->serviceFactory->createService($state)** which calls **$this->factory->make(Service::class, $state)** with $factory beeing FactoryInterface service injected at construct.

I don't like to be bound to a container implementation by I am until this is fixed. I hope there will be a time when container implementation will only be referenced in the application glue/bootstrap code.

A second idea would be that this interface is separating some concerns from ContainerInterface: creation of services. Basically, implementation would be a shorter version of the **get()**.

**Comment by glen-84** [comment]
Same question as [here](https://github.com/container-interop/container-interop/issues/43#issuecomment-183220525), namely, what is required to move this proposal forward?

@Ocramius @mnapoli -- could I set up a "survey" (series of questions) for the vendors to provide their input? Things like:

1. Should this be added.
2. To the main interface, or as an optional interface.
3. What should the method name be: A, B, C, D, other.
4. What should the method signature be: X, Y, Z, other.
5. Questions regarding behaviour.
6. Questions regarding behaviour when auto-wiring/resolution is enabled.
7. etc.

... or what is the usual process?

**Comment by pmjones** [comment]
Add to the list `newInstance($class, $params = [], $setters = [])` in [Aura.Di](https://github.com/auraphp/Aura.Di/blob/3.x/src/Container.php#L475-L510).

**Comment by mnapoli** [comment]
@glen-84 for #43 and #44 we need to see:

- if it's used by frameworks or library, i.e. will having an interface change anything?
- if PHP containers that have such a method today are willing to implement it (some won't, for example because container-interop is not a PSR, or because they don't want to rename/break BC, or because they don't see the point)

**Comment by glen-84** [comment]
> if it's used by frameworks or library, i.e. will having an interface change anything?

I don't know how I would find that information without searching through hundreds of libraries for multiple method signatures which could be unrelated to DI anyway. I'm not a library author; I want this for an application so that I can code to an interface instead of hard-coding for a specific implementation. I'm starting to give up on this ideal though.

> if PHP containers that have such a method today are willing to implement it

This would require the survey that I mentioned (or something similar).

**Comment by mnapoli** [comment]
Then you are starting to see how much work this project requires ;)

The point of interoperability is whether or not frameworks can be decoupled from container-specific features.

**Comment by glen-84** [comment]
> Then you are starting to see how much work this project requires ;)

I did offer to help. =)

> The point of interoperability is whether or not frameworks can be decoupled from container-specific features.

And this specific feature is implemented, in some way, by at least 5 containers. But if there is no interest, then I will just come up with a makeshift ugly-fix and try not to think about it too much.

**Comment by mnapoli** [comment]
There are 2 sides to an interface: implementors and consumers. Research is needed around frameworks that might use it.

**Comment by weierophinney** [comment]
I'll throw one more into the list: zend-servicemanager v3 uses `build()` (which was done on the recommendation of @mnapoli ) for plugin managers. Note that this is a separate context than the normal DI container; the assumption with `build()` is that you will *always* return a new instance, as the presence of instantiation options means it's acting as a *factory*. (Interestingly, `build()` also acts as an implementation detail, as we call it from `get()` if no cached instance is present yet.)

**Comment by mnapoli** [comment]
@weierophinney oh I didn't follow that it was actually merged, that's great!

By the way why did you make the type-hint for extra arguments `null` instead of an empty array? https://github.com/zendframework/zend-servicemanager/blob/master/src/ServiceLocatorInterface.php#L33

**Comment by mindplay-dk** [comment]
> zend-servicemanager v3 uses `build()` (which was done on the recommendation of @mnapoli ) for plugin managers. Note that this is a separate context than the normal DI container; the assumption with `build()` is that you will always return a new instance

@weierophinney I believe I have this feature in my DI container also?

https://github.com/mindplay-dk/unbox/blob/master/src/Container.php#L375

> Interestingly, `build()` also acts as an implementation detail, as we call it from `get()` if no cached instance is present yet

Appears to be the same approach I'm using :-)

**Comment by mnapoli** [comment]
@mindplay-dk yes you opened the issue and linked to that method ;)

**Comment by mindplay-dk** [comment]
@mnapoli hahaha, cripes, I'm doing to much stufffff.... ;-)

**Comment by weierophinney** [comment]
> why did you make the type-hint for extra arguments null instead of an empty array?

Honestly, I cannot remember exactly why; to hazard a guess, it was to be able to differentiate between no arguments passed (`null`) and arguments passed (`array`), and to allow empty arrays to be an indication of arguments being passed.

**Comment by glen-84** [comment]
So when can we start bikeshedding? :grin:

... I'm just kidding ... kind of.
